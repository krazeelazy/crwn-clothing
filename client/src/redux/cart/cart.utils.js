export const addItemToCart = (cartItems, cartItemToAdd) => {
    const existingCartItem = cartItems.find( // check if the item is already in the cart
        cartItem => cartItem.id === cartItemToAdd.id
    );
        
    if (existingCartItem) { // if the item is already in the cart use map to return a new array with the updated quantities
        return cartItems.map(cartItem =>
            cartItem.id === cartItemToAdd.id // find the item
            ? { ...cartItem, quantity: cartItem.quantity + 1 } // increment the quantity
            : cartItem // don't modify other items
            );
        }
        
        return [...cartItems, { ...cartItemToAdd, quantity: 1 }]; // if the item wasn't in the cart before, add it with a base quantity of 1
};

export const removeItemFromCart = (cartItems, cartItemToRemove) => {
    const existingCartItem = cartItems.find( // check for the item in the cart
        cartItem => cartItem.id === cartItemToRemove.id
    );

    if (existingCartItem.quantity === 1) { // if there's only 1 in the cart remove it
        return cartItems.filter(cartItem => cartItem.id !== cartItemToRemove.id);
    }

    // if the item is in the cart and the quantity is > 1 decrement the quantity by 1
    return cartItems.map(cartItem =>
        cartItem.id === cartItemToRemove.id
        ? { ...cartItem, quantity: cartItem.quantity - 1 }
        : cartItem
    );
};