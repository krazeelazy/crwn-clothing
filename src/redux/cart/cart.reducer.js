import CartActionTypes from './cart.types';

const INITIAL_STATE = {
    hidden: true, 
    cartItems : []
};

const cartReducer = (state = INITIAL_STATE, action) => {
    switch (action.type) {
        case CartActionTypes.TOGGLE_CART_HIDDEN:
            return {
                ...state,
                hidden: !state.hidden
            };
        case CartActionTypes.ADD_ITEM:
            return {
                ...state,
                cartItems: [...state.cartItems, action.payload] // just add the item to the array (will deal w/ duplicates later)
            };
        default:
            return state;
    }
};

export default cartReducer;