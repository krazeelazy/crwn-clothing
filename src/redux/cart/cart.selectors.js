import { createSelector } from 'reselect';

const selectCart = state => state.cart;// input selector

export const selectCartItems = createSelector( // a memoized selector
    [selectCart], // an array of input selectors (just 1 here)
    cart => cart.cartItems // a function that returns the value we want out of the selector
);

export const selectCartHidden = createSelector(
    [selectCart],
    cart => cart.hidden
);
    
export const selectCartItemsCount = createSelector(
    [selectCartItems],
    cartItems =>
        cartItems.reduce(
            (accumalatedQuantity, cartItem) =>
            accumalatedQuantity + cartItem.quantity,
            0
        )
);