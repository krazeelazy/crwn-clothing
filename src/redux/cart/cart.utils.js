export const addItemToCart = (cartItems, cartItemToAdd) => {
    const existingCartItem = cartItems.find( // check if the item is already in the cart
        cartItem => cartItem.id === cartItemToAdd.id
    );
        
    if (existingCartItem) { // if the item is already in the cart use map to return a new array with the updated quantities
        return cartItems.map(cartItem =>
            cartItem.id === cartItemToAdd.id // find the item
            ? { ...cartItem, quantity: cartItem.quantity + 1 } // increment the quantity
            : cartItem // don't modify other items
            );
        }
        
        return [...cartItems, { ...cartItemToAdd, quantity: 1 }]; // if the item wasn't in the cart before, add it with a base quantity of 1
};