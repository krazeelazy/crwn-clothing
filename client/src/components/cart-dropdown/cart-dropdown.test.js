import React from 'react';
import { shallow } from 'enzyme';
import * as redux from 'react-redux';

import CartDropdown  from './cart-dropdown.component';
import CartItem from '../cart-item/cart-item.component';

import { toggleCartHidden } from '../../redux/cart/cart.actions';

describe('CartDropdown component', () => {
    let wrapper;
    let spyOnUseSelector;
    let spyOnUseDispatch;
    let spyOnUseHistory;
    let mockDispatch;
    let mockHistory;
    const mockCartItems = [{ id: 1 }, { id: 2 }, { id: 3 }];

    beforeEach(() => {
        // Mock useSelector hook
        spyOnUseSelector = jest.spyOn(redux, 'useSelector');
        spyOnUseSelector.mockReturnValue({cartItems: mockCartItems});
    
        // Mock useDispatch hook
        spyOnUseDispatch = jest.spyOn(redux, 'useDispatch');
        // Mock dispatch function returned from useDispatch
        mockDispatch = jest.fn();
        spyOnUseDispatch.mockReturnValue(mockDispatch);

        // Mock history function returned from useHistory
        jest.mock('react-router-dom', () => ({
            useHistory: () => ({
                push: jest.fn(),
            }),
        }));

        console.log(mockHistory)

        wrapper = shallow(<CartDropdown />);
    });

    afterEach(() => {
        jest.restoreAllMocks();
    });

    it('should render CartDropdown component', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should call history.push when button is clicked', () => {
        wrapper.find('CartDropdownButton').simulate('click');
        expect(mockHistory.push).toHaveBeenCalled();
        expect(mockDispatch).toHaveBeenCalledWith(toggleCartHidden());
    });

    it('should render an equal number of CartItem components as the cartItems prop', () => {
        expect(wrapper.find(CartItem).length).toEqual(mockCartItems.length);
    });

    it('should render EmptyMessageContainer if cartItems is empty', () => {
        const newWrapper = shallow(<CartDropdown />);
        expect(newWrapper.exists('EmptyMessageContainer')).toBe(true);
    });
});