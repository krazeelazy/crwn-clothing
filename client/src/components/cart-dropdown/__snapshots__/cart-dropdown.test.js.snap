// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartDropdown component should render CartDropdown component 1`] = `
<styled.div>
  <styled.div>
    <EmptyMessageContainer>
      Your cart is empty
    </EmptyMessageContainer>
  </styled.div>
  <CartDropdownButton
    onClick={[Function]}
  >
    GO TO CHECKOUT
  </CartDropdownButton>
</styled.div>
`;
